src/QsCompiler/Compiler/CompilationLoader.cs

```diff
@@ -500,53 +500,55 @@ public CompilationLoader(SourceLoader loadSources, ReferenceLoader loadReference
+                steps.Add((rewriteStep.Priority, rewriteStep.Name, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.Monomorphization)));
             }
 
             if (this.config.GenerateFunctorSupport)
             {
                 var rewriteStep = new RewriteSteps.LoadedStep(new FunctorGeneration(), typeof(IRewriteStep), thisDllUri);
-                steps.Add((rewriteStep.Priority, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.FunctorSupport)));
+                steps.Add((rewriteStep.Priority, rewriteStep.Name, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.FunctorSupport)));
             }
 
             if (!this.config.SkipSyntaxTreeTrimming)
             {
                 var rewriteStep = new RewriteSteps.LoadedStep(new ConjugationInlining(), typeof(IRewriteStep), thisDllUri);
-                steps.Add((rewriteStep.Priority, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.TreeTrimming)));
+                steps.Add((rewriteStep.Priority, rewriteStep.Name, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.TreeTrimming)));
             }
 
             if (this.config.AttemptFullPreEvaluation)
             {
                 var rewriteStep = new RewriteSteps.LoadedStep(new FullPreEvaluation(), typeof(IRewriteStep), thisDllUri);
-                steps.Add((rewriteStep.Priority, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.PreEvaluation)));
+                steps.Add((rewriteStep.Priority, rewriteStep.Name, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.PreEvaluation)));
             }
 
             for (int j = 0; j < this.externalRewriteSteps.Length; j++)
             {
-                var priority = this.externalRewriteSteps[j].Priority;
+                var rewriteStep = this.externalRewriteSteps[j];
+                var priority = rewriteStep.Priority;
                 Func<QsCompilation> Execute(int index) => () =>
-                    this.ExecuteAsAtomicTransformation(this.externalRewriteSteps[index], ref this.compilationStatus.LoadedRewriteSteps[index]);
-                steps.Add((priority, Execute(j)));
+                    this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.LoadedRewriteSteps[index]);
+                steps.Add((priority, rewriteStep.Name, Execute(j)));
             }
 
-            PerformanceTracking.TaskStart(PerformanceTracking.Task.RewriteSteps);
             SortRewriteSteps(steps, t => t.Item1);
-            foreach (var (_, rewriteStep) in steps)
+            PerformanceTracking.TaskStart(PerformanceTracking.Task.RewriteSteps);
+            foreach (var (_, name, rewriteStep) in steps)
             {
+                PerformanceTracking.TaskStart(PerformanceTracking.Task.RewriteSteps, name);

```

Comment:
I would add a new item to the `PerformanceTracking.Task` enum for individual rewrite steps and use it here. Something like `IndividualRewriteStep` or `SingularRewriteStep`.

This way, every rewrite step would be hierarchically organized under the `RewriteSteps` task.

Remember to add the hierarchy specification in `TasksHierarchy`. Something like the following:
```csharp
{ Task.SingleRewriteStep, Task.RewriteSteps }
```

Reply:
I've added a separate enum for SingleRewriteStep, although I'd like to learn more about why this is necessary. I think there is something about this setup I'm not quite clear on. Let's chat offline about it.

Reply:
Discussed offline: Using an explicit enum value better supports the task hierarchy, as well as future work on this feature.